File: macos-app/Sources/ThoughtInput/Core/DestinationSender.swift
import Foundation

enum DestinationSender {
    static func send(
        payload: CapturePayload,
        destination: Destination,
        tokenManager: OAuthTokenManager
    ) async throws {
        let jsonData = try JSONEncoder().encode(payload)

        switch destination.type {
        case .supabase(let config):
            try await sendSupabase(data: jsonData, config: config)

        case .restNoAuth(let config):
            try await sendREST(data: jsonData, url: config.endpointURL, headers: [:])

        case .restApiKey(let config):
            let apiKey = try KeychainService.loadString(ref: config.apiKeyRef) ?? ""
            try await sendREST(data: jsonData, url: config.endpointURL, headers: [config.headerName: apiKey])

        case .restOAuthPassword, .restOAuthClientCredentials:
            try await sendWithOAuth(data: jsonData, destination: destination, tokenManager: tokenManager)
        }
    }

    // MARK: - Test Connection

    static func testConnection(destination: Destination, tokenManager: OAuthTokenManager) async throws {
        let testPayload = CapturePayload.create(text: "Connection test", method: .typed)
        try await send(payload: testPayload, destination: destination, tokenManager: tokenManager)
    }

    // MARK: - Supabase

    private static func sendSupabase(data: Data, config: SupabaseConfig) async throws {
        let urlString = config.projectURL
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            + "/rest/v1/" + config.tableName

        guard let url = URL(string: urlString) else {
            throw CaptureError.serverError(statusCode: -1)
        }

        let apiKey = try KeychainService.loadString(ref: config.apiKeyRef) ?? ""

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("return=minimal", forHTTPHeaderField: "Prefer")
        request.setValue(apiKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.httpBody = data
        request.timeoutInterval = 10

        try await execute(request)
    }

    // MARK: - REST

    private static func sendREST(data: Data, url: String, headers: [String: String]) async throws {
        guard let requestURL = URL(string: url) else {
            throw CaptureError.serverError(statusCode: -1)
        }

        var request = URLRequest(url: requestURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = data
        request.timeoutInterval = 10

        for (name, value) in headers {
            request.setValue(value, forHTTPHeaderField: name)
        }

        try await execute(request)
    }

    // MARK: - OAuth

    private static func sendWithOAuth(
        data: Data,
        destination: Destination,
        tokenManager: OAuthTokenManager
    ) async throws {
        let endpointURL: String
        switch destination.type {
        case .restOAuthPassword(let c): endpointURL = c.endpointURL
        case .restOAuthClientCredentials(let c): endpointURL = c.endpointURL
        default:
            CaptureLog.network.error("sendWithOAuth called for non-OAuth destination type")
            throw CaptureError.serverError(statusCode: -1)
        }

        let token = try await tokenManager.validToken(for: destination)

        do {
            try await sendREST(data: data, url: endpointURL, headers: ["Authorization": "Bearer \(token)"])
        } catch let error as CaptureError {
            // Retry once on 401
            if case .serverError(statusCode: 401) = error {
                await tokenManager.clearToken(for: destination.id)
                let newToken = try await tokenManager.validToken(for: destination)
                try await sendREST(data: data, url: endpointURL, headers: ["Authorization": "Bearer \(newToken)"])
            } else {
                throw error
            }
        }
    }

    // MARK: - Execute

    private static func execute(_ request: URLRequest) async throws {
        let (_, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw CaptureError.serverError(statusCode: code)
        }
    }
}

File: android-app/app/src/main/java/com/thoughtinput/capture/data/destinations/DestinationSender.kt
package com.thoughtinput.capture.data.destinations

import com.thoughtinput.capture.data.ApiClient
import com.thoughtinput.capture.data.CapturePayload

class CaptureHttpException(val statusCode: Int, val responseBody: String = "") :
    Exception("Server returned status $statusCode")

class DestinationSender(
    private val keystore: SecretsKeystore,
    private val tokenManager: OAuthTokenManager,
    private val apiClient: ApiClient = ApiClient()
) {

    suspend fun send(payload: CapturePayload, destination: Destination) {
        val data = payload.toJson().toByteArray(Charsets.UTF_8)
        when (val type = destination.type) {
            is DestinationType.Supabase -> sendSupabase(data, type)
            is DestinationType.RestNoAuth -> sendRest(data, type.endpointURL, emptyMap())
            is DestinationType.RestApiKey -> {
                val key = keystore.loadString(type.apiKeyRef).orEmpty()
                sendRest(data, type.endpointURL, mapOf(type.headerName to key))
            }
            is DestinationType.RestOAuthPassword,
            is DestinationType.RestOAuthClientCredentials -> sendWithOAuth(data, destination)
        }
    }

    suspend fun testConnection(destination: Destination) {
        val test = CapturePayload.create("Connection test", CapturePayload.CaptureMethod.TYPED)
        send(test, destination)
    }

    private suspend fun sendSupabase(data: ByteArray, config: DestinationType.Supabase) {
        val base = config.projectURL.trim().trimEnd('/')
        val url = "$base/rest/v1/${config.tableName}"
        val apiKey = keystore.loadString(config.apiKeyRef).orEmpty()
        val headers = mapOf(
            "Prefer" to "return=minimal",
            "apikey" to apiKey,
            "Authorization" to "Bearer $apiKey"
        )
        execute(url, headers, data)
    }

    private suspend fun sendRest(data: ByteArray, url: String, headers: Map<String, String>) {
        execute(url, headers, data)
    }

    private suspend fun sendWithOAuth(data: ByteArray, destination: Destination) {
        val endpointURL = when (val t = destination.type) {
            is DestinationType.RestOAuthPassword -> t.endpointURL
            is DestinationType.RestOAuthClientCredentials -> t.endpointURL
            else -> throw IllegalStateException("Non-OAuth type in sendWithOAuth")
        }

        val token = tokenManager.validToken(destination)
        try {
            execute(endpointURL, mapOf("Authorization" to "Bearer $token"), data)
        } catch (e: CaptureHttpException) {
            if (e.statusCode == 401) {
                tokenManager.clearToken(destination.id)
                val newToken = tokenManager.validToken(destination)
                execute(endpointURL, mapOf("Authorization" to "Bearer $newToken"), data)
            } else throw e
        }
    }

    private suspend fun execute(url: String, headers: Map<String, String>, data: ByteArray) {
        when (val r = apiClient.post(url, headers, data)) {
            is ApiClient.Result.Success -> Unit
            is ApiClient.Result.HttpError -> throw CaptureHttpException(r.statusCode, r.body)
            is ApiClient.Result.IoError -> throw CaptureHttpException(-1, r.message)
        }
    }
}
